runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -100,6 +103,40 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         val stream = conn.makeRequest(engineRequest, respHandler)
         stream.activate()
 
+        if (request.isChunked) {
+            val bytesToRead = 65536L
+
+            when (request.body) {
+                is HttpBody.SourceContent -> {
+                    val source = (request.body as HttpBody.SourceContent).readFrom()
+                    val nextBuffer = SdkBuffer()
+                    var buffer = SdkBuffer()
+
+                    while (true) {
+                        source.read(buffer, bytesToRead)
+
+                        // read another set of bytes into a second buffer -- this is how we can "peek" to decide if this is the final chunk or not
+                        val isFinalChunk = source.read(nextBuffer, bytesToRead) == -1L
+
+                        stream.writeChunk(buffer.readToByteArray(), isFinalChunk)
+
+                        if (isFinalChunk) break else buffer = nextBuffer
+                    }
```

Comment:
**Question**: Am I reading this correctly that the first `writeChunk` will get 64K, subsequent `writeChunk`s will get 128K (read, swap, read again), and the final `writeChunk` will get whatever's left? If so, does the initial chunk's smaller size matter?

Reply:
That's right, and the sizes are all arbitrary. We should select a size that we are comfortable loading into memory all-at-once, and I think 128K is fine.